galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker.go

```diff
@@ -0,0 +1,220 @@
+// fqdn and portNumber.
+func NewTargetServiceWithPortNumber(fqdn string, portNumber uint32) TargetService {
+	return TargetService{fQDN: fqdn, portNumber: portNumber}
+}
+
+// NewTargetServiceWithPortName creates a new TargetService using the specified fqdn
+// and portName.
+func NewTargetServiceWithPortName(fqdn, portName string) TargetService {
+	return TargetService{fQDN: fqdn, portName: portName}
+}
+
+// NewTargetService creates a new TargetService using the specified fqdn. Because no
+// port is specified, this implicitly represents the service bound to any port.
+func NewTargetService(fqdn string) TargetService {
+	return TargetService{fQDN: fqdn}
+}
+
+// FQDN is the fully-qualified domain name for the service (e.g.
+// foobar.my-namespace.svc.cluster.local).
+func (w TargetService) FQDN() string {
+	return w.fQDN
+}
+
+// PortNumber is the port used by the service. If set (non-zero), then
+// PortName must be the default value ("").
+func (w TargetService) PortNumber() uint32 {
+	return w.portNumber
+}
+
+// PortName is the name of the port used by the service. If set (not ""),
+// then PortNumber must be the default value (0).
+func (w TargetService) PortName() string {
+	return w.portName
+}
+
+// PolicyChecker allows callers to add a set of v1alpha1.Policy objects in the
+// mesh. Once these are loaded, you can query whether or not a specific
+// TargetService will require MTLS when an incoming connection occurs using the
+// IsServiceMTLSEnforced() call.
+type PolicyChecker struct {
+	// meshHasStrictMTLSPolicy tracks whether or not mTLS is strictly enforced on the mesh.
+	meshHasStrictMTLSPolicy bool
+
+	namespaceToMTLSMode map[string]strictMTLSMode
+	serviceToMTLSMode   map[TargetService]strictMTLSMode
+}
+
+// strictMTLSMode is a helper type used to represent whether or not MTLS was
+// explicitly enabled, explicitly disabled, or just not present. Useful for the
+// IsServiceMTLSEnforced check to determine which policy applies.
+type strictMTLSMode int
+
+const (
+	strictMTLSUnset strictMTLSMode = iota
+	strictMTLSExplicitlyEnabled
+	strictMTLSExplicitlyDisabled
+)
+
+// NewPolicyChecker creates a new PolicyChecker instance.
+func NewPolicyChecker() *PolicyChecker {
+	return &PolicyChecker{
+		namespaceToMTLSMode: make(map[string]strictMTLSMode),
+		serviceToMTLSMode:   make(map[TargetService]strictMTLSMode),
+	}
+}
+
+// AddMeshPolicy adds a mesh-level policy to the checker. Note that there can
+// only be at most one mesh level policy in effect.
+func (pc *PolicyChecker) AddMeshPolicy(p *v1alpha1.Policy) {
+	pc.meshHasStrictMTLSPolicy = doesPolicyEnforceMTLS(p)
+}
+
+// AddPolicy adds a new policy object to the PolicyChecker to use when later
+// determining if a service is MTLS-enforced. The namespace of the policy is
+// also provided as some policies can target the local namespace.
+func (pc *PolicyChecker) AddPolicy(namespace string, p *v1alpha1.Policy) error {
```

Comment:
I'm OK with this approach, but the mutability of PolicyChecker makes me a bit uncomfortable. I'm not seeing an Obviously Better alternative, though. We could pass all policy info to NewPolicyChecker, but we'd probably need a helper object to encapsulate [namespace,policy] tuples. We could also use a builder pattern, so AddPolicy becomes WithPolicy.

Not hugely important, but it bothers me mildly.

Reply:
I'm glad you feel this way - it's part of the reason I added so many tests and pulled the code into its own package. The mutability is extremely uncomfortable but also the most direct way to solve it (and ultimately mirrors how pilot works underneath).

If it helps, mutable structs are not my default approach to solving problems like this, despite my background in Java :D